[게시글 생성/조회/수정/삭제] 이바 미션 제출합니다.

seonghun127 opened the pull request:
이번 미션은 CRUD 기능 구현까지 진행했습니다.
서버 API를 구현하는데 ATDD 방식으로 진행하려고 노력했습니다.
ATDD라는 것이 아직 정확히 몰라서 그냥 TDD로 진행했다고 보는게 더 맞을 것 같습니다.
API에 대한 테스트 코드 작성이 많이 어려웠습니다.
테스트 코드를 비롯한 여러 부분에서 부족한 점이 많을테니 신랄하게 지적해주시면 감사하겠습니다.

seonghun127 committed:
setting: mission complete

seonghun127 committed:
docs: README.md

1. todo 목록 정리

seonghun127 committed:
feat: build.gradle

1. devtools 의존성 추가

seonghun127 committed:
feat: ArticleController Class

1. Article 도메인 객체 구현
2. Test 추가
3. ArticleRepository 저장기능 구현

seonghun127 committed:
feat: layout

1. build.gradle 의존성 추가
2. 공통 레이아웃 추출 및 적용

seonghun127 committed:
feat: ArticleDto Class

1. Create 용도의 ArticleDto 구현
2. ArticleDto를 사용하도록 Controller와 Repository 변경

seonghun127 committed:
refactor: Article class

1. ArticleDto를 사용하기 위한 생성자 추가
2. id 자동 증가를 위한 static field 추가

seonghun127 committed:
feat: ArticleController class

1. 조회 기능 구현

seonghun127 committed:
style: css 및 templates 파일 코드 포메팅

seonghun127 committed:
fix: app.css

1. 파일 오류 수정

seonghun127 committed:
test: ArticleRepository class

1. 조회 및 전체조회 테스트

seonghun127 committed:
fix: layout

1. thymeleaf-layout 버전에 따른 오류 수정

seonghun127 committed:
fix: ArticleController class

1. createArticle 리다이렉팅 주소 수정

seonghun127 committed:
feat: ArticleController class

1. 게시물 수정 기능 구현
2. ArticleDto.Update 생성
3. ArticleRepository에 update 기능 추가
4. update를 위해 Article 도메인 객체에 생성자 추가

seonghun127 committed:
feat: article-edit.html

1. 생성 화면과 수정 화면에 따른 form 경로 및 메서드 조정
2. article.html에 하드코딩된 url 변경

seonghun127 committed:
feat: ArticleController class

1. 삭제 기능 구현
2. ArticleRepository delete 구현

seonghun127 committed:
feat: article.html

1. delete 버튼 구현

seonghun127 committed:
refactor: layout

1. header 공통부분 추출

seonghun127 committed:
fix: ArticleRepository class

1. 수정시 새 게시물이 추가되는 오류 수정

seonghun127 committed:
style: 전체 코드 포매팅

seonghun127 committed:
fix: article-edit.html

1. 수정 페이지에서 글 내용이 나오지 않는 오류 수정

seonghun127 committed:
fix: article.html

1. 예시로 설정된 게시글 내용 삭제
2. 작성된 게시글 내용이 나오게끔 수정

seonghun127 committed:
refactor: custom head 적용 방법 변경

seonghun127 committed:
refactor: fragment 폴더 생성

seonghun127 committed:
refactor: layout.html

1. main-head를 담는 th:replace 태그 추가
2. main-body를 담는 th:replace 태그 추가

seonghun127 committed:
refactor: ArticleController class

1. 필드 의존 주입에서 생성자 의존 주입으로 변경

seonghun127 committed:
refactor: NotFoundArticleException class 추가

seonghun127 committed:
refactor: ArticleRepository class

1. find() 메서드의 이름 수정 : findById()

seonghun127 committed:
style: 코드 포매팅

kang-hyungu reviewed (changes requested):
안녕하세요 이바님 리뷰어 강현구입니다.
field injection을 사용하지 않고 컨트롤러를 구현하셨네요 :+1: 
몇 가지 피드백 남겼으니 확인해주세요!

seonghun127 committed:
refactor: Article class

1. set 메서드 제거

seonghun127 committed:
fix: ArticleRepositoryTest 클래스에서 깨지는 테스트 코드 수정

seonghun127 committed:
refactor: ArticleController class

1. @AllArgsContructor 를 @RequiredArgsContructor 어노테이션으로 변경
2. articleRepository 필드를 final 변경
- 불변셩을 유지하기 위해

seonghun127 committed:
feat: Article class

1. 커스터마이즈한 Builder 패턴 추가

seonghun127 committed:
refactor: Article class에 lombok @Builder 적용

seonghun127 committed:
feat: ArticleController class

1. 게시물 추가 테스트 코드 구현

seonghun127 committed:
style: ArticleRepositoryTests class 이름 변경

seonghun127 committed:
fix: Article class

1. updateArticle() 메서드 추가

seonghun127 committed:
refactor: 중복 head 부분 제거

seonghun127 committed:
feat: ArticleController class

1. modelMapper 의존성 주입
2. modelMapper 클래스를 빈으로 등록
3. ArticleDto 클래스에 Response 중첩클래스 추가
4. ArticleController에서 article(s) 를 반환하는 모든 API의 반환 형태를 ArticleDto.Response 타입으로 변경

seonghun127 committed:
style: 코드 포맷팅

kang-hyungu reviewed (changes requested):
반영하신 내용 확인했습니다!
질문 주신 내용 답변이랑 article controller test에 추가 코멘트 남긴 부분 확인 부탁드려요!
수정하시고 머지 요청 주시면 확인하는대로 바로 반영할게요

seonghun127 committed:
refactor: ArticleController class

1. updateArticle() 메서드 리턴 방식 변경
- 리다이렉트에서 포워딩 방식으로 변경한다
2. 그에 따른 인수 테스트 방식 수정
- 수정된 값을 확인하는 테스트로 수정한다

seonghun127 committed:
style: 주석 제거

seonghun127 committed:
refactor: ArticleRespository class

1. save() 메서드 리턴 타입 void로 변경
2. update() 메서드 리턴 타입 void로 변경

seonghun127 committed:
refactor: ArticleController class

1. createArticle() 메서드 리턴 타입 변경
- 리다이렉트 방식에서 포워딩 방식으로 변경한다

seonghun127 committed:
refactor: ArticleControllerTests class

1. 지속 테스트 가능한 코드로 리펙토링

seonghun127 committed:
style: 코드 포맷팅

seonghun127 commented:
`ArticleControllerTests`에서 `ArticleRepository` 인스턴스를 사용하지않는 방식으로 리펙토링을 진행했습니다.
게시물에 대한 CRUD 테스트를 수행하면서 지속적으로 증가될 `Article` 도메인의 `id`를 관리하기 위해 테스트 클래스 내부에 `static int currentArticleId`로 뒀습니다.
그리고 삭제 테스트 메서드를 `@AfterEach`로 변경하였습니다. `@AfterEach` 가 수행될때마다 `currentArticleId`를 1씩 증가시켜줬습니다.

테스트 코드에서 리다이렉트되는 게시물 생성, 수정 관련 api에 대해 새롭게 생성(수정)되는 값을 확인하기 위해 포워드 방식으로 api를 변경했습니다. (리다이렉트되는 api에서는 불가능하다고 생각했었습니다;;)
하지만 생성, 수정관련 기능에 대해 포워드 방식이 이뤄질 경우 요청시 생성되는 request 객체의 정보가 그대로 남아있어 새로고침 시 request 객체에 남아있는 정보대로 다시 서버와 통신이 이뤄져 같은 글이 계속 생성되거나 의도치 않은 수정이 이뤄질 수 있다는 문제를 알게되었습니다. (리뷰 요청을 보내고 나서야 알게되어...)
부랴부랴 해당 api를 리다이렉트 방식으로 수정하고 테스트 코드에서 리다이렉트되는 api에 대한 테스트 코드를 작성하는 법에 찾아보고 알맞게 수정했습니다.

seonghun127 committed:
fix: ArticleController class

1. createArticle(), updateArticle() api 리턴 타입 변경
- 포워드 방식에서 리다이렉트 방식으로 변경한다

2. ArticleRepository 클래스에서 save(), update() 메서드의 리턴 타입 변경
- void에서 long으로 변경한다

3. 테스트 코드 수정

seonghun127 committed:
style: 코드 포맷팅

kang-hyungu reviewed (approved):
수정하신 내용 확인했어요!
코드 구현하느라 수고하셨어요 👏
